components/range_cache_memory_engine/src/engine.rs

```diff
@@ -336,38 +317,69 @@ impl RangeCacheMemoryEngine {
 
     // It handles the pending range and check whether to buffer write for this
     // range.
-    pub(crate) fn prepare_for_apply(
-        &self,
-        write_batch_id: u64,
-        region: &CacheRegion,
-    ) -> RangeCacheStatus {
-        let mut core = self.core.write();
-        let range_manager = core.mut_range_manager();
-        let Some(mut region_state) = range_manager.check_region_state(region) else {
+    pub(crate) fn prepare_for_apply(&self, region: &CacheRegion) -> RangeCacheStatus {
+        let manager = self.core.region_manager();
+        if !manager.is_active() {
+            return RangeCacheStatus::NotInCache;
+        }
+
+        // fast path, only need to hold the read lock.
+        {
+            let regions_map = manager.regions_map.read();
+            let Some(region_meta) = regions_map.region_meta(region.id) else {
+                return RangeCacheStatus::NotInCache;
+            };
+            let state = region_meta.get_state();
+            if state == RegionState::Active {
+                region_meta.set_being_written();
+                return RangeCacheStatus::Cached;
+            } else if state.is_evict() {
+                return RangeCacheStatus::NotInCache;
+            } else if state == RegionState::Loading {
+                region_meta.set_being_written();
+                return RangeCacheStatus::Loading;
+            }
+        }
+
+        // slow path, handle pending region
+        let mut regions_map = manager.regions_map.write();
+        let cached_count = regions_map.regions().len();
+        let Some(region_meta) = regions_map.mut_region_meta(region.id) else {
```

Comment:
When will this return None?

Reply:
It's possible that the region is evicted before acquiring the write lock again(after release the read lock in the first block), thus it can be None.